packages/nice-grpc-client-middleware-devtools/src/index.ts

```diff
@@ -0,0 +1,216 @@
+    // skip streaming calls
+    if (call.requestStream || call.responseStream) {
+      return yield* call.next(call.request, options);
+    }
+
+    // log unary calls
+    const { path } = call.method;
+    const reqObj = getAsObject(call.request);
+
+    try {
+      const result = yield* call.next(call.request, options);
+      const resObj = getAsObject(result);
+      window.postMessage(
+        {
+          method: path,
+          methodType: 'unary',
+          request: reqObj,
+          response: resObj,
+          type: '__GRPCWEB_DEVTOOLS__',
+        },
+        '*'
+      );
+      return result;
+    } catch (error) {
+      if (error instanceof ClientError) {
+        window.postMessage(
+          {
+            error: {
+              code: error?.code,
+              message: `${error?.message || error}`,
+              name: error?.name,
+              stack: error?.stack,
+            },
+            method: path,
+            methodType: 'unary',
+            request: reqObj,
+            type: '__GRPCWEB_DEVTOOLS__',
+          },
+          '*'
+        );
+      } else if (isAbortError(error) && error instanceof Error) {
+        if (!options.skipAbortErrorLogging) {
+          window.postMessage(
+            {
+              error: {
+                code: 1,
+                message: `${error?.message || error}`,
+                name: error?.name,
+                stack: error?.stack,
+              },
+              method: path,
+              methodType: 'unary',
+              request: reqObj,
+              type: '__GRPCWEB_DEVTOOLS__',
+            },
+            '*'
+          );
+        }
+      } else if (error instanceof Error) {
+        window.postMessage(
+          {
+            error: {
+              code: 2,
+              message: `${error?.message || error}`,
+              name: error?.name,
+              stack: error?.stack,
+            },
+            method: path,
+            methodType: 'unary',
+            request: reqObj,
+            type: '__GRPCWEB_DEVTOOLS__',
+          },
+          '*'
+        );
+      }
+
+      throw error;
+    }
+  };
+
+export const devtoolsStreamLoggingMiddleware: ClientMiddleware<DevtoolsLoggingOptions> =
+  async function* devtoolsLoggingMiddleware<Request, Response>(
+    call: ClientMiddlewareCall<Request, Response>,
+    options: CallOptions & Partial<DevtoolsLoggingOptions>
+  ): AsyncGenerator<Response, Response | void, undefined> {
+    // skip unary calls
+    if (!call.responseStream && !call.requestStream) {
+      return yield* call.next(call.request, options);
+    }
+
+    // log streaming calls
+    const { path } = call.method;
+    const reqObj = getAsObject(call.request);
+
+    let first = true;
+    try {
+      for await (const response of call.next(call.request, options)) {
+        const resObj = getAsObject(response);
+        if (first) {
+          // log the request object only once
+          window.postMessage(
```

Comment:
It seems to me that we can log the request object before even sending the request. For example if the stream does not ever send a response, we should still be able to see the request in the dev tools.

Also consider that the request can be a stream too, in which case `call.request` would be an `AsyncIterable`. Not sure whether the dev tools support client streams though.

Reply:
The logic to log the request only once the first response is received is by intention.
This way if there's an error during the call there will be exactly one message in dev tools indicating that error. Logging the request before the call is executed would result in two messages in dev tools: One for the intended request and an additional one for the error.

For the client-streaming and bidirectional streaming I've updated the implemented, though it's not prominently used in grpc-web I think...